### ISSUE TITLE
[Feature] Add Model Export to ONNX Format

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ProjectOdyssey#2670]
## Problem

Models trained in ML Odyssey can't be easily deployed or used in other frameworks without ONNX export.

### 1. Basic ONNX Exporter

```mojo
# shared/export/onnx.mojo
struct ONNXExporter:
    """Export Mojo models to ONNX format."""
    var opset_version: Int
    
    fn __init__(out self, opset_version: Int = 14):
        self.opset_version = opset_version
    
    fn export(self, model: Model, sample_input: ExTensor, output_path: String):
        """Export model to ONNX format.
        
        Args:
            model: Model to export
            sample_input: Example input for shape inference
            output_path: Path to save ONNX file
        """
        # Trace model execution
        var graph = self._trace_model(model, sample_input)
        
        # Convert to ONNX graph
        var onnx_graph = self._convert_to_onnx(graph)
        
        # Write ONNX protobuf
        write_onnx_file(output_path, onnx_graph, self.opset_version)
        
        print("Model exported to:", output_path)
```

### 2. Operation Mapping

```mojo
fn _convert_operation(op: Operation) -> ONNXNode:
    """Convert ML Odyssey operation to ONNX node."""
    if op.type == "Linear":
        return ONNXNode("Gemm", op.inputs, op.outputs, op.attributes)
    
    elif op.type == "Conv2d":
        return ONNXNode("Conv", op.inputs, op.outputs, {
            "kernel_shape": op.kernel_size,
            "strides": op.stride,
            "pads": op.padding,
        })
    
    elif op.type == "ReLU":
        return ONNXNode("Relu", op.inputs, op.outputs, {})
    
    elif op.type == "MaxPool2d":
        return ONNXNode("MaxPool", op.inputs, op.outputs, {
            "kernel_shape": op.kernel_size,
            "strides": op.stride,
        })
    
    # ... more operations
```

### 3. Usage Example

```mojo
fn export_lenet5():
    """Export trained LeNet5 to ONNX."""
    # Load trained model
    var model = LeNet5()
    model.load_state_dict("checkpoints/best_model.bin")
    model.eval()  # Set to inference mode
    
    # Create sample input (for shape inference)
    var sample_input = ExTensor([1, 1, 28, 28], DType.float32)
    
    # Export to ONNX
    var exporter = ONNXExporter(opset_version=14)
    exporter.export(model, sample_input, "lenet5.onnx")
    
    print("Model exported successfully!")
    print("Use with: onnxruntime, TensorRT, OpenVINO, etc.")
```

### 4. Verification

```python
# Verify ONNX model (Python script)
import onnx
import onnxruntime as ort
import numpy as np

# Load and check ONNX model
model = onnx.load("lenet5.onnx")
onnx.checker.check_model(model)
print("ONNX model is valid!")

# Run inference with ONNX Runtime
session = ort.InferenceSession("lenet5.onnx")
input_name = session.get_inputs()[0].name
output_name = session.get_outputs()[0].name

# Test input
x = np.random.randn(1, 1, 28, 28).astype(np.float32)

# Run inference
output = session.run([output_name], {input_name: x})
print("Output shape:", output[0].shape)
```

### 5. Advanced: Dynamic Axes

```mojo
fn export_with_dynamic_batch():
    """Export model with dynamic batch size."""
    var model = LeNet5()
    var sample_input = ExTensor([1, 1, 28, 28], DType.float32)
    
    var exporter = ONNXExporter()
    exporter.export(
        model,
        sample_input,
        "lenet5_dynamic.onnx",
        dynamic_axes={
            "input": {0: "batch_size"},  # Batch dimension is dynamic
            "output": {0: "batch_size"}
        }
    )
```

### 6. Metadata and Documentation

```mojo
fn export_with_metadata():
    """Export model with metadata."""
    var exporter = ONNXExporter()
    exporter.set_metadata({
        "model_name": "LeNet5",
        "description": "MNIST digit classifier",
        "author": "ML Odyssey",
        "version": "1.0",
        "license": "MIT"
    })
    
    exporter.export(model, sample_input, "lenet5_annotated.onnx")
```

## Use Cases

**Deployment:**
- ONNX Runtime (CPU/GPU inference)
- TensorRT (NVIDIA GPU optimization)
- OpenVINO (Intel hardware)
- CoreML (Apple devices)

**Model Analysis:**
- Netron (visualization)
- ONNX Simplifier (optimization)
- ONNX Shape Inference

**Interoperability:**
- Share models with PyTorch/TensorFlow users
- Deploy to production without Mojo runtime

## Supported Operations

Should support all common operations:
- Linear, Conv2d, BatchNorm2d
- ReLU, Sigmoid, Tanh, Softmax
- MaxPool2d, AvgPool2d
- Reshape, Transpose, Concat

## Impact

- **Severity**: LOW
- **Deployment**: Enable production deployment
- **Interoperability**: Share models across frameworks

**Comment by mvillmow** [comment]
## Dependencies (AI-Analyzed)

**Priority**: P2

### Depends On
- #2638

### Blocks
- #2675

### Analysis
ONNX export depends on checkpoint/serialization system (#2638). Nice-to-have feature for deployment but doesn't block core training. Supports PyTorch migration guide.

---
*Analyzed by Claude Code dependency analyzer*
